include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+    inputs_.push_back(producer->results_.GetChecked(output_index));
+    Succeed(producer);
+    return this;
+  }
+
+  /** Returns a value returned by one of the producers.
+   *
+   * @param index   The index of the corresponding call to Consume; NOT the output index.
+   *                consumer->Consume(producer, 1);
+   *                consumer->Consume(producer, 42);
+   *
+   *                GetInputValue(0)  // gets producer's output 1
+   *                GetInputValue(1)  // gets producer's output 42
+   */
+  const std::any &GetInputValue(int index) const {
+    if (state != TaskState::Running)
+      throw std::logic_error(
+          "Obtaining a value of a task's input is only valid inside a task's payload function.");
+    if (index < 0 || static_cast<size_t>(index) >= inputs_.size())
+      throw std::out_of_range("The specified input index is out of range.");
+    return inputs_[index]->Value();
+  }
+
+  /** Returns a value returns by one of the producers and casts it to the specified type.
+   */
+  template <typename T>
+  T GetInputValue(int index) const {
+    const std::any &result = GetInputValue(index);
+    if constexpr (!std::is_void_v<T>)
+      return std::any_cast<T>(inputs_[index]->Value());
+  }
+
+  /** Ensures that the releasable object is released after the task completes.
+   *
+   * The releasable object will be released once the task is complete - even if it fails with
+   * an exception.
+   */
+  Task *ReleaseAfterRun(std::shared_ptr<Releasable> releasable) {
+    if (state != TaskState::New)
+      throw std::logic_error(
+          "Cannot add a new postcondition to a task that's been submitted to execution.\n"
+          "If you need to Release a releasable object after completion of an already "
+          "submitted task, create an auxiliary dependent task.");
+    release_.push_back(std::move(releasable));
+    return this;
+  }
+
+  /**  Guards the execution of the task with a waitable/releasable object.
+   *
+   * Equivalent to Succeed + ReleaseAfterRun
+   */
+  Task *GuardWith(std::shared_ptr<Releasable> releasable) {
+    Succeed(releasable);
+    ReleaseAfterRun(std::move(releasable));
+    return this;
+  }
+
+  /**  Executes the task. The task must have been submitted to the specified scheduler.
+   */
+  void Run(Scheduler &sched);
+
+
+ private:
+  SharedTask next;        // pointer to the next task in an intrusive list
+  Task *prev = nullptr;   // pointer to the previous task in an intrusive list
+
+  friend class detail::TaskList;
+  friend class Scheduler;
+
+  TaskResults results_;
+
+  double priority_ = 0;
+  std::function<void(Task *)> wrapped_;
+  SmallVector<std::shared_ptr<Waitable>, 4> preconditions_;
+  SmallVector<SharedTaskResult, 4> inputs_;
+  SmallVector<std::shared_ptr<Releasable>, 4> release_;
+};
+
+namespace detail {
+
+/** An intrusive list of tasks
+ *
+ * This list uses tasks' built-in next and prev fields.
```

Comment:
Can you describe a convention in this list. What is the head, tail, next and prev.

Reply:
I thought these were the most common names used in bidirectional lists...

Reply:
Sometimes head can point to itself if it is empty (we have an empty root that you cannot remove), or just be empty. That is what I mean.